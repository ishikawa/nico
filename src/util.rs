use std::cell::RefCell;
use std::rc::Rc;

pub fn wrap<T>(ty: T) -> Rc<RefCell<T>> {
    Rc::new(RefCell::new(ty))
}

pub trait UniqueNaming {
    fn next(&mut self) -> String;
}

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug)]
pub struct SequenceNaming {
    type_var_index: usize,
    prefix: String,
}

impl SequenceNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }
}

impl UniqueNaming for SequenceNaming {
    fn next(&mut self) -> String {
        let i = self.type_var_index;
        self.type_var_index += 1;
        format!("{}{}", self.prefix, i)
    }
}
