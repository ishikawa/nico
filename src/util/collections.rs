use crate::arena::BumpaloVec;
use std::cell::Ref;

#[derive(Debug)]
pub struct RefVecIter<'r, 'arena, T> {
    vec: Ref<'r, BumpaloVec<'arena, T>>,
}

impl<'r, 'arena, T> RefVecIter<'r, 'arena, T> {
    pub fn new(vec: Ref<'r, BumpaloVec<'arena, T>>) -> Self {
        Self { vec }
    }
}

impl<'a: 'r, 'r, 'arena, T> IntoIterator for &'a RefVecIter<'r, 'arena, T> {
    type IntoIter = std::slice::Iter<'r, T>;
    type Item = &'r T;

    fn into_iter(self) -> Self::IntoIter {
        self.vec.iter()
    }
}
