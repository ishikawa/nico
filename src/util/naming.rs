use std::collections::HashMap;

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug, Default)]
pub struct SequenceNaming {
    names: HashMap<String, i32>,
}

impl SequenceNaming {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next<S: AsRef<str>>(&mut self, name: S) -> String {
        let name = name.as_ref();
        let n = self.names.entry(name.to_string()).or_insert(0);
        let next = format!("{}.{}", name, n);

        *n += 1;
        next
    }
}

/// This struct produces a unique name while giving a sequential
/// number to the name passed to the next() method.
#[derive(Debug)]
pub struct PrefixNaming {
    type_var_index: usize,
    prefix: String,
}

impl PrefixNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }

    pub fn next(&mut self) -> String {
        let i = self.type_var_index;
        let next = format!("{}{}", self.prefix, i);

        self.type_var_index += 1;
        next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prefix_naming() {
        let mut naming = PrefixNaming::new("my_");

        assert_eq!(naming.next(), "my_0".to_string());
        assert_eq!(naming.next(), "my_1".to_string());
    }

    #[test]
    fn sequence_naming() {
        let mut naming = SequenceNaming::new();

        assert_eq!(naming.next("x"), "x.0".to_string());
        assert_eq!(naming.next("x"), "x.1".to_string());
        assert_eq!(naming.next("y"), "y.0".to_string());
    }
}
