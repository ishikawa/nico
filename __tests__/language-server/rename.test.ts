import { LanguageServer, spawn_server, LanguageServerAgent, Position } from "../util/lsp";
import { filterTestCases, TestCaseBase, readTestFileSync, getTestName } from "../util/testcase";

interface TestCase extends TestCaseBase {
  requests: Array<{ position: Position; newName: string }>;
}

let server: LanguageServer | undefined;

beforeAll(async () => {
  server = await spawn_server({ rename: true });
});

afterAll(() => {
  if (server) {
    return server.stop();
  }
});

let cases: TestCase[] = [
  // function name
  {
    // prettier-ignore
    input: [
      "fun bar()",
      "    100",
      "end",
      "",
      "bar()"
    ].join("\n"),
    requests: [
      {
        position: {
          line: 4,
          character: 3
        },
        newName: "bar"
      }
    ]
  },
  // struct name
  {
    // prettier-ignore
    input: [
      "struct A { b: i32 }",
      "",
      "let a = A { b: 123 }",
      "a.b",
    ].join("\n"),
    requests: [
      {
        position: {
          line: 0,
          character: 7
        },
        newName: "B"
      }
    ]
  },
  // function parameter
  {
    // prettier-ignore
    input: [
      "fun foo(x)",
      "    x",
      "end",
    ].join("\n"),
    requests: [
      // param: x
      {
        position: {
          line: 0,
          character: 8
        },
        newName: "bar"
      },
      // variable: x
      {
        position: {
          line: 1,
          character: 4
        },
        newName: "foo"
      }
    ]
  },
  // let variables
  {
    // prettier-ignore
    input: [
      "let foo = 100",
      "println_i32(foo)",
    ].join("\n"),
    requests: [
      // let foo = 100
      {
        position: {
          line: 0,
          character: 4
        },
        newName: "foo"
      },
      // println_i32(foo)
      {
        position: {
          line: 1,
          character: 12
        },
        newName: "foo"
      }
    ]
  },
  // variable name
  {
    // prettier-ignore
    input: [
      "struct A { b: i32 }",
      "",
      "let a = A { b: 123 }",
      "a.b"
    ].join("\n"),
    requests: [
      // let a
      {
        position: {
          line: 2,
          character: 4
        },
        newName: "c"
      },
      // a.b
      {
        position: {
          line: 3,
          character: 0
        },
        newName: "foo"
      }
    ]
  },
  // array pattern
  {
    // prettier-ignore
    input: [
      "let x = [1, 2, 3]",

      "case x",
      "when [x, y, z]",
      "    x + y + z",
      "else",
      "    100",
      "end",
    ].join("\n"),
    requests: [
      // when [x, y, z]
      //       ^
      {
        position: {
          line: 3,
          character: 6
        },
        newName: "a"
      },
      // when [a, y, z]
      //          ^
      {
        position: {
          line: 3,
          character: 9
        },
        newName: "b"
      },
      // when [a, b, z]
      //             ^
      {
        position: {
          line: 3,
          character: 12
        },
        newName: "c"
      },
      //     a + b + c
      //     ^
      {
        position: {
          line: 4,
          character: 4
        },
        newName: "x"
      },
      //     a + b + c
      //         ^
      {
        position: {
          line: 4,
          character: 8
        },
        newName: "y"
      },
      //     a + b + c
      //             ^
      {
        position: {
          line: 4,
          character: 13
        },
        newName: "z"
      }
    ]
  },
  // struct field
  {
    // prettier-ignore
    input: [
      "struct A { b: i32 }",
      "",
      "let a = A { b: 123 }",
      "a.b"
    ].join("\n"),
    requests: [
      // struct A { b: i32 }
      //            ^
      {
        position: {
          line: 0,
          character: 11
        },
        newName: "c"
      },
      // let a = A { b: 123 }
      //             ^
      {
        position: {
          line: 2,
          character: 12
        },
        newName: "c"
      },
      // a.b
      //   ^
      {
        position: {
          line: 3,
          character: 2
        },
        newName: "c"
      }
    ]
  }
];

filterTestCases(cases).forEach((testCase, i) => {
  let src = readTestFileSync(testCase);
  let name = getTestName(testCase);

  testCase.requests.forEach(({ position, newName }, j) => {
    test(`${i}: rename - \`${name}\` at ${position}`, async done => {
      const agent = new LanguageServerAgent(server!, { sequence: i * 100 + j });

      // Open document and no compilation errors
      const diagnostics = await agent.openDocument(name, src);
      expect(diagnostics).toHaveLength(0);

      const response = await agent.rename(name, position, newName);
      expect(response).toMatchSnapshot();

      done();
    });
  });
});
