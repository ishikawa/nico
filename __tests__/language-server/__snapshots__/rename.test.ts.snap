// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0: prepare rename - \`fun bar()
    100
end

bar()\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1000,
  "jsonrpc": "2.0",
  "result": Object {
    "changes": Object {
      "file:///home/user/nico/samplefun%20bar()%0A%20%20%20%20100%0Aend%0A%0Abar().nico": Array [
        Object {
          "newText": "bar",
          "range": Object {
            "end": Object {
              "character": 7,
              "line": 0,
            },
            "start": Object {
              "character": 4,
              "line": 0,
            },
          },
        },
        Object {
          "newText": "bar",
          "range": Object {
            "end": Object {
              "character": 3,
              "line": 4,
            },
            "start": Object {
              "character": 0,
              "line": 4,
            },
          },
        },
      ],
    },
  },
}
`;

exports[`1: prepare rename - \`struct A { b: i32 }

let a = A { b: 123 }
a.b\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1100,
  "jsonrpc": "2.0",
  "result": Object {
    "changes": Object {
      "file:///home/user/nico/samplestruct%20A%20%7B%20b%3A%20i32%20%7D%0A%0Alet%20a%20%3D%20A%20%7B.nico": Array [
        Object {
          "newText": "B",
          "range": Object {
            "end": Object {
              "character": 8,
              "line": 0,
            },
            "start": Object {
              "character": 7,
              "line": 0,
            },
          },
        },
        Object {
          "newText": "B",
          "range": Object {
            "end": Object {
              "character": 9,
              "line": 2,
            },
            "start": Object {
              "character": 8,
              "line": 2,
            },
          },
        },
      ],
    },
  },
}
`;

exports[`2: prepare rename - \`fun foo(x)
    x
end\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1200,
  "jsonrpc": "2.0",
  "result": Object {
    "changes": Object {
      "file:///home/user/nico/samplefun%20foo(x)%0A%20%20%20%20x%0Aend.nico": Array [
        Object {
          "newText": "bar",
          "range": Object {
            "end": Object {
              "character": 9,
              "line": 0,
            },
            "start": Object {
              "character": 8,
              "line": 0,
            },
          },
        },
        Object {
          "newText": "bar",
          "range": Object {
            "end": Object {
              "character": 5,
              "line": 1,
            },
            "start": Object {
              "character": 4,
              "line": 1,
            },
          },
        },
      ],
    },
  },
}
`;

exports[`2: prepare rename - \`fun foo(x)
    x
end\` at [object Object] 2`] = `
Object {
  "error": null,
  "id": 1201,
  "jsonrpc": "2.0",
  "result": Object {
    "changes": Object {
      "file:///home/user/nico/samplefun%20foo(x)%0A%20%20%20%20x%0Aend.nico": Array [
        Object {
          "newText": "foo",
          "range": Object {
            "end": Object {
              "character": 9,
              "line": 0,
            },
            "start": Object {
              "character": 8,
              "line": 0,
            },
          },
        },
        Object {
          "newText": "foo",
          "range": Object {
            "end": Object {
              "character": 5,
              "line": 1,
            },
            "start": Object {
              "character": 4,
              "line": 1,
            },
          },
        },
      ],
    },
  },
}
`;

exports[`3: prepare rename - \`struct A { b: i32 }

let a = A { b: 123 }
a.b\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1300,
  "jsonrpc": "2.0",
  "result": Object {
    "changes": Object {
      "file:///home/user/nico/samplestruct%20A%20%7B%20b%3A%20i32%20%7D%0A%0Alet%20a%20%3D%20A%20%7B.nico": Array [
        Object {
          "newText": "c",
          "range": Object {
            "end": Object {
              "character": 5,
              "line": 2,
            },
            "start": Object {
              "character": 4,
              "line": 2,
            },
          },
        },
        Object {
          "newText": "c",
          "range": Object {
            "end": Object {
              "character": 1,
              "line": 3,
            },
            "start": Object {
              "character": 0,
              "line": 3,
            },
          },
        },
      ],
    },
  },
}
`;

exports[`3: prepare rename - \`struct A { b: i32 }

let a = A { b: 123 }
a.b\` at [object Object] 2`] = `
Object {
  "error": null,
  "id": 1301,
  "jsonrpc": "2.0",
  "result": Object {
    "changes": Object {
      "file:///home/user/nico/samplestruct%20A%20%7B%20b%3A%20i32%20%7D%0A%0Alet%20a%20%3D%20A%20%7B.nico": Array [
        Object {
          "newText": "foo",
          "range": Object {
            "end": Object {
              "character": 5,
              "line": 2,
            },
            "start": Object {
              "character": 4,
              "line": 2,
            },
          },
        },
        Object {
          "newText": "foo",
          "range": Object {
            "end": Object {
              "character": 1,
              "line": 3,
            },
            "start": Object {
              "character": 0,
              "line": 3,
            },
          },
        },
      ],
    },
  },
}
`;
